contracts/BC_fusion/StakeCredit.sol

```diff
@@ -0,0 +1,346 @@
+    // delegator address => personal unbond sequence
+    mapping(address => CountersUpgradeable.Counter) private _unbondSequence;
+
+    // day index => receivedReward
+    mapping(uint256 => uint256) public rewardRecord;
+    // day index => totalPooledBNB
+    mapping(uint256 => uint256) public totalPooledBNBRecord;
+
+    /*----------------- structs and events -----------------*/
+    struct UnbondRequest {
+        uint256 shares;
+        uint256 bnbAmount;
+        uint256 unlockTime;
+    }
+
+    event RewardReceived(uint256 rewardToAll, uint256 commission);
+
+    /**
+     * @notice only accept BNB from `StakeHub`
+     */
+    receive() external payable onlyStakeHub {
+        uint256 index = block.timestamp / IStakeHub(STAKE_HUB_ADDR).BREATHE_BLOCK_INTERVAL();
+        totalPooledBNBRecord[index] = totalPooledBNB;
+        rewardRecord[index] += msg.value;
+        totalPooledBNB += msg.value;
+    }
+
+    /*----------------- init -----------------*/
+    /*
+     * @param _validator validator's operator address
+     * @param _moniker validator's moniker
+     */
+    function initialize(address _validator, string calldata _moniker) external payable initializer onlyStakeHub {
+        string memory name_ = string.concat("Stake ", _moniker, " Credit");
+        string memory symbol_ = string.concat("st", _moniker);
+        __ERC20_init_unchained(name_, symbol_);
+        __ReentrancyGuard_init_unchained();
+
+        validator = _validator;
+
+        _bootstrapInitialHolder(msg.value);
+    }
+
+    /*----------------- external functions -----------------*/
+    /**
+     * @param delegator the address of the delegator
+     * @return shares the amount of shares minted
+     */
+    function delegate(address delegator) external payable onlyStakeHub returns (uint256 shares) {
+        if (msg.value == 0) revert ZeroAmount();
+        shares = _mintAndSync(delegator, msg.value);
+        if (shares == 0) revert ZeroShares();
+    }
+
+    /**
+     * @param delegator the address of the delegator
+     * @param shares the amount of shares to be undelegated
+     * @return bnbAmount the amount of BNB to be unlocked
+     */
+    function undelegate(address delegator, uint256 shares) external onlyStakeHub returns (uint256 bnbAmount) {
+        if (shares == 0) revert ZeroShares();
+        if (shares > balanceOf(delegator)) revert InsufficientBalance();
+
+        // add to the queue
+        bnbAmount = _burnAndSync(delegator, shares);
+        uint256 unlockTime = block.timestamp + IStakeHub(STAKE_HUB_ADDR).unbondPeriod();
+        UnbondRequest memory request = UnbondRequest({ shares: shares, bnbAmount: bnbAmount, unlockTime: unlockTime });
+        bytes32 hash = keccak256(abi.encodePacked(delegator, _useSequence(delegator)));
+        // the hash should not exist in the queue
+        // this will not happen in normal cases
+        if (_unbondRequests[hash].shares != 0) revert RequestExisted();
```

Comment:
undelegate same shares in the same block will call this, right? so we forbiden such cases.

Reply:
No, undelegate same shares in the same block will not have this problem because `_useSequence` will increase the user's nonce by 1 everytime.